src/librustc/traits/project.rs

```diff
@@ -1177,7 +1179,7 @@ fn confirm_callable_candidate<'cx, 'gcx, 'tcx>(
     // Note: we unwrap the binder here but re-create it below (1)
     let ty::Binder((trait_ref, ret_type)) =
         tcx.closure_trait_ref_and_return_type(fn_once_def_id,
-                                              obligation.predicate.trait_ref.self_ty(),
+                                              obligation.predicate.trait_ref(selcx.tcx()).self_ty(),
```

Comment:
Don't need to call `trait_ref` to get something from the substs. Maybe add `self_ty` to `ProjectionTy`.